apps/openmw/mwphysics/physicssystem.cpp

```diff
@@ -739,6 +740,18 @@ namespace MWPhysics
         delete mBroadphase;
     }
 
+    void PhysicsSystem::updateIdle()
+    {
+        for (ActorMap::iterator it = mActors.begin(); it != mActors.end(); ++it)
+        {
+            osg::Vec3f pos(it->second->getCollisionObjectPosition());
+
+            RayResult result = castRay(pos, pos - osg::Vec3f(0, 0, it->second->getHalfExtents().z() + 2), it->second->getPtr(), std::vector<MWWorld::Ptr>(), CollisionType_World|CollisionType_HeightMap|CollisionType_Door);
+
+            it->second->setIdle(result.mHit);
+        }
```

Comment:
This function has a lot of problems:
1) It's unnecessary to check actors out of physics processing range.
2) Even with 1) fixed, an extra raytest for every actor every frame is too heavy performance wise. I'm sure you would notice a difference in the profiling graph.
3) A ray test is not very accurate for this. If the actor is standing on a slope, the center of its collision base may be elevated and falsely detected as not on ground. A convex cast using the actor's collision shape would be more accurate, like in the code that updates onGround status in PhysicsSystem::move. Then again, I have no idea why we would want to duplicate that code in the first place.

I think there's a much easier way to achieve what you want. In MovementSolver::move, make it so sGroundOffset is not added to the movement result if the actor is not moving. Physics can stay enabled that way and you can delete all the code you added except for the adjustPosition changes.

Reply:
> an extra raytest for every actor every frame is too heavy performance wise.

Why every frame? An engine should update it only twice per second. Or I did something wrong?

> It's unnecessary to check actors out of physics processing range.

Got it.

> A ray test is not very accurate for this. If the actor is standing on a slope, the center of its collision base may be elevated and falsely detected as not on ground. A convex cast using the actor's collision shape would be more accurate, like in the code that updates onGround status in PhysicsSystem::move. Then again, I have no idea why we would want to duplicate that code in the first place.

Actually, we do not need very accurate results here: there are cases, when an object (e.g. chair) inside actor's collision box. In this case result would be unpredictable.

If something really went wrong, there is a Fall console command now.